src/read.c

```diff
@@ -1248,146 +1393,58 @@ void ReadRecExpr (
 **  'ReadFuncExpr' reads a function literal expression.  In  case of an error
 **  it skips all symbols up to one contained in <follow>.
 **
-**  <Function> := 'function (' [ <Ident> {',' <Ident>} ] ')'
+**  <Function> := 'function (' <ArgList> ')'
 **                             [ 'local'  <Ident> {',' <Ident>} ';' ]
 **                             <Statments>
 **                'end'
 */
-
-
 void ReadFuncExpr (
     TypSymbolSet        follow,
     Char mode)
 {
-    volatile Obj        nams;           /* list of local variables names   */
     volatile Obj        name;           /* one local variable name         */
-    volatile Int        narg;           /* number of arguments             */
-    volatile UInt       isvarg = 0;     /* does function have varargs?     */
-    volatile UInt       nloc;           /* number of locals                */
     volatile UInt       nr;             /* number of statements            */
     volatile UInt       i;              /* loop variable                   */
     volatile UInt       nrError;        /* copy of <TLS(NrError)>          */
     volatile Bag        currLVars;      /* copy of <TLS(CurrLVars)>             */
     volatile Int        startLine;      /* line number of function keyword */
     volatile int        is_block = 0;   /* is this a do ... od block?      */
     volatile int        is_atomic = 0;  /* is this an atomic function?      */
-    volatile int        lockmode;       /* type of lock for current argument */
-    volatile Bag        locks = 0;      /* locks of the function */
+    volatile Int        narg;           /* number of arguments             */
+    volatile Obj        nams;           /* list of local variables names   */
+    volatile UInt       nloc;           /* number of locals                */
+    UInt                isvarg = 0;     /* is this function variadic?      */
+    nloc = 0;
 
     /* begin the function               */
     startLine = TLS(Input)->number;
     if (TLS(Symbol) == S_DO) {
 	Match( S_DO, "do", follow );
         is_block = 1;
+        /* make and push the new local variables list (args and locals)        */
+        narg = 0;
+        nams = NEW_PLIST( T_PLIST, narg );
+        SET_LEN_PLIST( nams, narg );
+        TLS(CountNams) += 1;
+        ASS_LIST( TLS(StackNams), TLS(CountNams), nams );
```

Comment:
In GAP we disabled the parsing for `do od` blocks, so I'm fairly sure this can never get called, but obviously haven't removed all the associated code.